### ISSUE TITLE
Installation issue: boost

### ISSUE THREAD

**Issue by payerle** [LLNL/spack#31237]
### Steps to reproduce the issue

```console
$ spack spec -I boost
Input spec
--------------------------------
boost

Concretized
--------------------------------
jahffyi  boost@1.72.0%gcc@8.4.0+atomic+chrono~clanglibcpp+container+context+contract+coroutine+date_time~debug+exception+fiber+filesystem+graph+graph_parallel~icu+iostreams+json+locale+log+math~mpi+multithreaded+nowide~numpy+pic+program_options~python+random+regex+serialization+shared+signals~singlethreaded+stacktrace+system~taggedlayout+test+thread+timer+type_erasure~versionedlayout+wave context-impl=fcontext cxxstd=11 patches=5da7ad2,e13cca1 visibility=hidden arch=linux-rhel8-zen
ntqbk4r      ^bzip2@1.0.8%gcc@8.4.0~debug+pic+shared arch=linux-rhel8-zen
xt7llpb          ^diffutils@3.8%gcc@8.4.0 arch=linux-rhel8-zen
pzvi3uo              ^libiconv@1.16%gcc@8.4.0 libs=shared,static arch=linux-rhel8-zen
utajygi      ^zlib@1.2.12%gcc@8.4.0+optimize+pic+shared patches=0d38234 arch=linux-rhel8-zen
```

### Error message

<details><summary>Error message</summary><pre>==> boost: Executing phase: 'install'
==> [2022-06-22-12:01:32.693438] './bootstrap.sh' '--prefix=/cvmfs/hpcsw.umd.edu/spack-software/2022.06.15/linux-rhel8-zen/gcc-8.4.0/boost-1.72.0-jahffyijqqtsivirjwh34v7gfjy7j23c' '--with-toolset=gcc' '--with-libraries=atomic,chrono,container,context,contract,coroutine,date_time,exception,fiber,filesystem,graph,graph_parallel,iostreams,locale,log,math,nowide,program_options,random,regex,serialization,stacktrace,system,test,thread,timer,type_erasure,wave' '--without-icu'
Building Boost.Build engine with toolset gcc... tools/build/src/engine/b2
Unicode/ICU support for Boost.Regex?... disabled.
Generating Boost.Build configuration in project-config.jam for gcc...

Bootstrapping is done. To build, run:

    ./b2

To generate header files, run:

    ./b2 headers

To adjust configuration, edit 'project-config.jam'.
Further information:

   - Command line help:
     ./b2 --help

   - Getting started guide:
     http://www.boost.org/more/getting_started/unix-variants.html

   - Boost.Build documentation:
     http://www.boost.org/build/

==> [2022-06-22-12:01:39.037709] FILTER FILE: /tmp/devspackswinst/spack-stage/spack-stage-boost-1.72.0-jahffyijqqtsivirjwh34v7gfjy7j23c/spack-src/project-config.jam [replacing "^\s*using gcc.*"]
==> [2022-06-22-12:01:39.042658] './b2' '--clean' '-j' '16' '--user-config=/tmp/devspackswinst/spack-stage/spack-stage-boost-1.72.0-jahffyijqqtsivirjwh34v7gfjy7j23c/spack-src/user-config.jam' 'variant=release' '--disable-icu' '-s' 'BZIP2_INCLUDE=/cvmfs/hpcsw.umd.edu/spack-software/2022.06.15/linux-rhel8-zen/gcc-8.4.0/bzip2-1.0.8-ntqbk4r4tosrcs7rdf37qucmzy6fdcbl/include' '-s' 'BZIP2_LIBPATH=/cvmfs/hpcsw.umd.edu/spack-software/2022.06.15/linux-rhel8-zen/gcc-8.4.0/bzip2-1.0.8-ntqbk4r4tosrcs7rdf37qucmzy6fdcbl/lib' '-s' 'ZLIB_INCLUDE=/cvmfs/hpcsw.umd.edu/spack-software/2022.06.15/linux-rhel8-zen/gcc-8.4.0/zlib-1.2.12-utajygisfk2k4ly6omwczvmmykj66xkq/include' '-s' 'ZLIB_LIBPATH=/cvmfs/hpcsw.umd.edu/spack-software/2022.06.15/linux-rhel8-zen/gcc-8.4.0/zlib-1.2.12-utajygisfk2k4ly6omwczvmmykj66xkq/lib' '-s' 'NO_LZMA=1' '-s' 'NO_ZSTD=1' 'context-impl=fcontext' 'link=static,shared' '--layout=system' 'toolset=gcc' 'cxxstd=11' 'cxxflags="-fPIC"' 'visibility=hidden'
error: wrong library name 'nowide' in the --with-<library> option.
</pre></details>

### Information on your system

```console
spack debug report
* **Spack:** 0.19.0.dev0 (0205fefe0c80dc1fe03b9ec5da93bf24b1c9103f)
* **Python:** 3.6.8
* **Platform:** linux-rhel8-zen2
* **Concretizer:** clingo
```

### Additional information

[spack-build-out.txt](https://github.com/spack/spack/files/8959626/spack-build-out.txt)
[spack-build-env.txt](https://github.com/spack/spack/files/8959640/spack-build-env.txt)

Maintainers: @hainest 

Basically, I am building an old version of boost which predates support for the json and nowide libraries (seem to have been introduced in 1.75.0 and 1.73.0, respectively).  Either spack recipe should conflict, or as done in other cases in boost, remove libraries not supported by the version.

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by hainest** [comment]
@payerle Were you trying to install boost as a dependency through a package or just via `spack install boost@...`?

**Comment by payerle** [comment]
Sorry, I withdrew the PR but neglected to close the ticket. 
The approach given in #30627 is better than my suggestion
